### ISSUE TITLE
Possible regression when specifying duplicate paths

### ISSUE THREAD

**Issue by johanbrandhorst** [grpc-ecosystem/grpc-gateway#830]
I haven't got a lot of time to debug this right now, but at work I've recently run into several instances of the generator _not_ erroring when specifying duplicate paths, like so:

```
get: "/api/v1/users/{user_id}"
...
get: "/api/v1/users/things"
```

The second one should not be allowed because it is ambiguous with the first one. I'm pretty sure this used to work, so I assume it's a regression. I'm raising just so I don't forget about it. This is happening on latest master.

**Comment by srenatus** [comment]
I was thinking that the code generator should catch these -- but really, we could generate a bunch of `.pb.gw.go` files for different services, and end up using them together. Assuming that, we'd have a hard time catching this overlap at "compile time", but rather would check at "runtime". The latter could achieved by having `runtime.NewPattern` check any new one against the existing ones.

However, I could also see benefit in catching overlaps at compile time for those situations where we could. 🤔 

WDYT? Runtime or compile time?

**Comment by srenatus** [comment]
Guess we could make [`(*mux).Handle()`](https://github.com/grpc-ecosystem/grpc-gateway/blob/ae4bb97e7718691c8e0742a7c07989c91630e042/runtime/mux.go#L142-L145) return an error when a pattern clash is detected...

**Comment by johanbrandhorst** [comment]
I don't think we should need to generate several files or worry about runtime errors - the generator should catch this at generation time.

**Comment by srenatus** [comment]
Ok, that will only catch half the trouble, though. You could still have 
```
get: "/api/v1/users/{user_id}"
```
in `a.proto`, and
```
get: "/api/v1/users/things"
```
in `b.proto`; generate both `a.pb.gw.go` and `b.pb.gw.go` separately, and have a surprise in waiting when using the service with these two loaded. Or am I missing something?

**Comment by johanbrandhorst** [comment]
Hm, well yeah if the generator doesn't have access to both files that'd be impossible, but I suppose you should really be specifying all proto files that are in the same namespace in one `protoc` invocation?

**Comment by srenatus** [comment]
Looking at the pattern code, it would be nice to come up with a function like
```go
func (p Pattern) Overlaps(q Pattern) bool
```
that could then be used; but frankly, it's quite complicated, and I haven't reached a satisfying solution yet. (I make no claims on this issue whatsoever. Up for grabs.)

**Comment by lukaszx0** [comment]
I've just hit this issue for two endpoints *within* the same `service`, so this seems to be not covered even for the simple case. I'd expect compiler to fail at compile-time. It was very confusing to debug at first as we were getting marshalling errors - the runtime tried to unmarshall into wrong request proto struct.
